File: 2023-02/python/part1.py
import re

MAX_CUBES = dict(red=12, green=13, blue=14)
CUBE_REGEX = re.compile(r"(\d+) (\w+)")


def is_game_possible(game_string):
    return all((MAX_CUBES[c] >= int(n) for n, c in CUBE_REGEX.findall(game_string)))


if __name__ == "__main__":
    with open("../input.txt", 'r') as f:
        print(sum((i for i, line in enumerate(f.readlines(), 1) if is_game_possible(line))))

File: 2023-02/rust/part1.rs
use regex::Regex;
use std::collections::HashMap;
use std::fs::read_to_string;

fn run(filename: &str) -> usize {
    let max_cubes = HashMap::from([("red", 12), ("green", 13), ("blue", 14)]);
    let cube_regex = Regex::new(r"(\d+) (\w+)").unwrap();

    read_to_string(filename)
        .unwrap()
        .lines()
        .enumerate()
        .filter(|(_, line)| {
            cube_regex
                .captures_iter(line)
                .all(|c| max_cubes[&c[2]] >= c[1].parse::<i32>().unwrap())
        })
        .map(|(idx, _)| idx + 1)
        .sum()
}

fn main() {
    let result = run("../input.txt");
    println!("{result}");
}
